File: week-1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

// validate card number using luhn's algorithm
bool validate_checksum(long number);
// count the number of digits in the card number
int count_digits(long number);
// determine the card type based on length and prefix
void check_card_type(long number);

int main(void)
{
    // get the credit card number from the user
    long card_number = get_long("Number: ");

    // check if the card number passes the checksum
    if (validate_checksum(card_number))
    {
        // determine and print the card type
        check_card_type(card_number);
    }
    else
    {
        // print invalid if checksum fails
        printf("INVALID\n");
    }

    return 0;
}

bool validate_checksum(long number)
{
    int sum = 0;
    bool multiply = false;

    // apply luhn's algorithm
    while (number > 0)
    {
        int digit = number % 10;

        if (multiply)
        {
            // multiply every other digit by 2 and sum the digits of the product
            int product = digit * 2;
            sum += (product / 10) + (product % 10);
        }
        else
        {
            // add digits not multiplied by 2
            sum += digit;
        }

        // toggle the multiply flag
        multiply = !multiply;
        // move to the next digit
        number /= 10;
    }

    // return true if the total modulo 10 is 0
    return (sum % 10 == 0);
}

int count_digits(long number)
{
    int count = 0;

    // count the number of digits in the number
    while (number > 0)
    {
        number /= 10;
        count++;
    }

    return count;
}

void check_card_type(long number)
{
    // get the number of digits in the card number
    int digits = count_digits(number);

    // extract the first two digits of the card number
    long prefix = number;
    while (prefix >= 100)
    {
        prefix /= 10;
    }

    // check for american express
    if ((digits == 15) && (prefix == 34 || prefix == 37))
    {
        printf("AMEX\n");
    }
    // check for mastercard
    else if ((digits == 16) && (prefix >= 51 && prefix <= 55))
    {
        printf("MASTERCARD\n");
    }
    // check for visa
    else if ((digits == 13 || digits == 16) && (prefix / 10 == 4))
    {
        printf("VISA\n");
    }
    // if no match, print invalid
    else
    {
        printf("INVALID\n");
    }
}

File: week-6/sentimental-credit/credit.py
from enum import Enum


class CardType(Enum):
    VISA = "VISA"
    MASTERCARD = "MASTERCARD"
    AMEX = "AMEX"
    INVALID = "INVALID"


def main():
    # get card number from user with validation
    while True:
        try:
            number = int(input("Number: "))
            if number > 0:
                break
        except ValueError:
            print(CardType.INVALID.value)

    # check if valid card number and get type
    if validate_checksum(number):
        print(get_card_type(number).value)
    else:
        print(CardType.INVALID.value)


def validate_checksum(number):
    total = 0
    position = 0

    # implement luhn's algorithm
    while number > 0:
        digit = number % 10

        # multiply every other digit by 2 starting from second-to-last
        if position % 2 == 1:
            digit *= 2
            if digit > 9:
                # sum digits if product is greater than 9
                digit = digit // 10 + digit % 10
        total += digit

        number //= 10
        position += 1

    return total % 10 == 0


def get_card_type(number):
    num_str = str(number)
    length = len(num_str)

    # check visa first (starts with 4)
    if length in [13, 16] and num_str[0] == "4":
        return CardType.VISA

    # check mastercard (starts with 51-55)
    if length == 16 and num_str[:2] in ["51", "52", "53", "54", "55"]:
        return CardType.MASTERCARD

    # check amex (starts with 34 or 37)
    if length == 15 and num_str[:2] in ["34", "37"]:
        return CardType.AMEX

    # if no match found
    return CardType.INVALID


if __name__ == "__main__":
    main()
